### ISSUE TITLE
Warn when adding an optional parameter to a public method

### ISSUE THREAD

**Issue by chucker** [dotnet/roslyn#21782]
**Version Used**: 
15.3.2

**Steps to Reproduce**:

1. In an assembly `A`, create a public method `MyMethod()` without an optional parameter
2. Reference this assembly from `B`
3. Add an optional parameter (with a default value) to `MyMethod()`

In my particular case, a library had a class `DBSettings` with this member:

```csharp
public static bool TryGet<T>(string key, out T value);
```

I added a parameter as follows:

```csharp
public static bool TryGet<T>(string key, out T value, DBSettingValueScope scope = DBSettingValueScope.Any);
```

Previously, the scope was *always* `Any`. Now, the consumer of this API can optionally specify a different scope.

**Expected Behavior**:

This change doesn't appear to alter the public contract, and I'm not getting warned about any problems, so there shouldn't be any side effects. At compile time, there are no apparent problems.

**Actual Behavior**:

During run time, I'm getting a missing method exception in assembly `B`:

```
System.MissingMethodException: 'Method not found: 'Boolean AssemblyA.DBSettings.TryGet(System.String, !!0 ByRef)'.'
```

The runtime should either be smart enough to find the new method with the optional parameter, *or* the IDE should warn me that the public contract has in fact been altered.

**Comment by Pilchie** [comment]
Closing due to lack of info - please re-open with more information if this is still an issue for you.
